lib/globus/client/endpoint.rb

```diff
@@ -61,22 +57,33 @@ def connection
         )
       end
 
+      # Builds up a path from a list of path elements. E.g., input would look like:
+      #     ["mjgiarlo", "work123", "version1"]
+      # And this method returns:
+      #     ["/uploads/mjgiarlo/", "/uploads/mjgiarlo/work123/", "/uploads/mjgiarlo/work123/version1/"]
+      def paths
+        path_segments.map.with_index do |_segment, index|
+          File.join(config.uploads_directory, path_segments.slice(..index)).concat("/")
+        end
+      end
+
+      def path_segments
+        [user_id, "work#{work_id}", "version#{work_version}"]
+      end
```

Comment:
This is the new iteration-based strategy for building up a directory tree. I added method documentation above to make it more approachable.